test/Reader/Http/ResponseTest.php

```diff
@@ -0,0 +1,175 @@
+            'X-Content-Length' => 1234.56,
+        ]);
+        $this->assertEquals(204, $response->getStatusCode());
+        $this->assertEquals('', $response->getBody());
+        $this->assertEquals('http://example.org/foo', $response->getHeaderLine('Location'));
+        $this->assertEquals(1234, $response->getHeaderLine('Content-Length'));
+        $this->assertEquals(1234.56, $response->getHeaderLine('X-Content-Length'));
+    }
+
+    public function invalidStatusCodes()
+    {
+        foreach (range(-100, 99) as $statusCode) {
+            yield $statusCode => [$statusCode, 'between 100 and 599'];
+        }
+
+        foreach (range(600, 1000, 10) as $statusCode) {
+            yield $statusCode => [$statusCode, 'between 100 and 599'];
+        }
+
+        foreach ([100.1, 599.1] as $statusCode) {
+            yield $statusCode => [$statusCode, 'integer status code'];
+        }
+
+        $invalidTypes = [
+            'null'   => [null, 'numeric status code'],
+            'true'   => [true, 'numeric status code'],
+            'false'  => [false, 'numeric status code'],
+            'string' => [' 100 ', 'numeric status code'],
+            'array'  => [[200], 'numeric status code'],
+            'object' => [(object) [100], 'numeric status code'],
+        ];
+        foreach ($invalidTypes as $key => $value) {
+            yield $key => $value;
+        }
+    }
+
+    /**
+     * @dataProvider invalidStatusCodes
+     */
+    public function testConstructorRaisesExceptionForInvalidStatusCode($statusCode, $contains)
+    {
+        $this->setExpectedException(InvalidArgumentException::class, $contains);
+        new Response($statusCode);
+    }
+
+    public function invalidBodies()
+    {
+        return [
+            'null'       => [null],
+            'true'       => [true],
+            'false'      => [false],
+            'zero'       => [0],
+            'int'        => [1],
+            'zero-float' => [0.0],
+            'float'      => [1.1],
+            'array'      => [['BODY']],
+            'object'     => [(object) ['body' => 'BODY']],
+        ];
+    }
+
+    /**
+     * @dataProvider invalidBodies
+     */
+    public function testConstructorRaisesExceptionForInvalidBody($body)
+    {
+        $this->setExpectedException(InvalidArgumentException::class);
+        new Response(200, $body);
+    }
+
+    public function invalidHeaders()
+    {
+        return [
+            'empty-name' => [
+                [ '' => 'value' ],
+                'non-empty, non-numeric',
+            ],
+            'zero-name' => [
+                [ 'value' ],
+                'non-empty, non-numeric',
+            ],
+            'int-name' => [
+                [ 1 => 'value' ],
+                'non-empty, non-numeric',
+            ],
+            'numeric-name' => [
+                [ '1.1' => 'value' ],
+                'non-empty, non-numeric',
+            ],
+            'null-value'       => [
```

Comment:
> I think this data provider could has been represented as columns (one line for each data set)

I found it harder to understand what the arguments were when they were on a single line (i.e., where the array of headers ended and the message began). I'm going to leave them as-is.

> Also could this provider to be extracted in a common trait for reuse with ZendHttpClientDecoratorTest?

No; they're not equivalent usages or contexts.